### PR TITLE
feat: Adding RealGRU module and associated functions

### PR DESCRIPTION
**Pull request by MarIniOnz**
Added the RealGRU module, which is used to calculate the hidden states for the real data in the dataset.

### CODE REVIEW

File: medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            train_loader (MTGANDataLoader): The data loader for the training data.
+
+        Returns:
+            float: The loss of the model.
+        """
+        torch.manual_seed(0)
+        loss = torch.Tensor([np.inf])
+
+        for _ in range(1, self.epochs + 1):
+            for _, data in enumerate(train_loader, start=1):
```
Comment (JabobKrauskopf) [inline medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py:112]: why enumerate starting with 1 if you throw the index away anyways?

**Review by JabobKrauskopf — CHANGES_REQUESTED** [review]

File: medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+        Returns:
+            torch.Tensor: hidden state of the RealGRU with respect to the data, masked
+        """
+        with torch.no_grad():
+            maximum_number_of_windows = data.shape[1]
+            mask = find_sequence_mask(
+                number_of_windows_per_patient, maximum_number_of_windows
+            ).unsqueeze(dim=-1)
+            outputs, _ = self.gru(data)
```
Comment (LauraBoenchenLB) [inline medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py:96]: Aren't the hidden states on the second position of the GRU output?

File: medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py (hunk 1 above)
Comment (JabobKrauskopf) [reply medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py:112]: also the line above is the same as `for _ in range(0, self.epochs)`, no? (if the output is ignored via `_`)

Comment (LauraBoenchenLB) [inline medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py:112]: why use enumerate if we don't want the count? :D

**Review by LauraBoenchenLB — CHANGES_REQUESTED** [review]
Very cool functions, the GRU is looking goood! Some comments and questions

File: medmodels/data_synthesis/mtgan/model/masks.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    sequence and False for positions that are outside the length of the sequence.
+
+    Example:
+        number_windows = torch.tensor([1,3,2,2])
+        maximum_number_of_windows = 3
+        sequence_mask(number_windows, maximum_number_of_windows)
+        tensor([[[ True], # 1 > 0
+            [False], # 1 !> 1
+            [False]], # 1 !> 2
+
```
[the remark this answers is not in the document]
Comment (MarIniOnz) [reply medmodels/data_synthesis/mtgan/model/masks.py:20]: Let me see if this convinced you now

File: medmodels/data_synthesis/mtgan/model/real_gru/loss.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        self,
+        input_data: torch.Tensor,
+        predictions: torch.Tensor,
+        number_of_windows_per_patient: torch.Tensor,
+    ) -> torch.Tensor:
+        """Run forward method with masked input: compute binary cross entropy.
+
+        Args:
+            input_data (torch.Tensor): Output of the GRU.
+            predictions (torch.Tensor): boolean matrix of shape (batch size, maximum number of
```
[the remark this answers is not in the document]
Comment (MarIniOnz) [reply medmodels/data_synthesis/mtgan/model/real_gru/loss.py:40]: Explained better now in the docstrings. Let me know if it still not clear enough.

File: medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py (hunk 2 above)
Comment (MarIniOnz) [reply medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py:96]: The first position is the hidden states at each time step for each sequence in the batch. Here we are aiming to work with these sequential hidden states (h_t).

The second position corresponds to the hidden states ONLY from the last timestep of each sequence (h_n).

File: medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    optimizer: torch.optim.optimizer.Optimizer
+    loss_function: nn.Module
+
+    def __init__(
+        self,
+        total_number_of_concepts: int,
+        hyperparamaters: TrainingHyperparameters,
+        device: torch.device,
+    ) -> None:
+        """Constructor fot the RealGRU.
```
Comment (LauraBoenchenLB) [inline medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py:34]:
```suggestion
        """Constructor for the RealGRU.
```

**Review by LauraBoenchenLB — CHANGES_REQUESTED** [review]
Great changes, docstrings are clearer now :) Found a few little typos, approval is near!

**Review by JabobKrauskopf — APPROVED** [review]
LGTM!

**Review by LauraBoenchenLB — APPROVED** [review]
AMAZING 🦭 
IMG_1380_Original
